app/src/test/java/com/github/se/cyrcle/model/user/UserRepositoryFirestoreTest.kt

```diff
@@ -55,92 +70,212 @@ class UserRepositoryFirestoreTest {
+  fun getUidTest_throwsException_whenUidIsEmpty() {
+    `when`(mockFirebaseAuth.currentUser).thenReturn(null)
+    try {
+      userRepositoryFirestore.getUid()
+    } catch (e: Exception) {
+      assert(e.message == "User not signed in")
+    }
+  }
 
-    // Ensure that mockParkingQuerySnapshot is properly initialized and mocked
-    `when`(mockCollectionReference.get()).thenReturn(taskCompletionSource.task)
+  @Test
+  fun onSignIn_callsOnSuccess() {
+    `when`(mockFirebaseAuth.currentUser).thenReturn(mockFirebaseUser)
+    userRepositoryFirestore.onSignIn { assert(mockFirebaseAuth.currentUser != null) }
+  }
 
-    // Ensure the QuerySnapshot returns a list of mock DocumentSnapshots
-    `when`(mockParkingQuerySnapshot.documents).thenReturn(listOf(mockDocumentSnapshot))
+  @Test
+  fun getUserById_callsOnSuccess() {
+    // Create a list of QueryDocumentSnapshots
+    val queryDocumentSnapshots = listOf(mockQueryDocumentSnapshot)
 
-    // Ensure that the DocumentSnapshot returns the expected Parking object
+    `when`(mockCollectionReference.get()).thenReturn(Tasks.forResult(mockUserQuerySnapshot))
+    `when`(mockDocumentReference.get()).thenReturn(Tasks.forResult(mockDocumentSnapshot))
+    `when`(mockUserQuerySnapshot.documents).thenReturn(queryDocumentSnapshots)
     `when`(mockDocumentSnapshot.toObject(User::class.java)).thenReturn(user)
 
-    // Call the method to be tested
-    userRepositoryFirestore.getAllUsers(
-        onSuccess = { users ->
-          assert(users.size == 1)
-          assert(users[0] == user)
-        },
+    userRepositoryFirestore.getUserById(
+        userId = user.public.userId,
+        onSuccess = { assert(true) },
         onFailure = { fail("Expected success but got failure") })
-    // Manually complete the task
-    taskCompletionSource.setResult(mockParkingQuerySnapshot)
 
-    // Verify that the method was called
-    verify(timeout(100)) { (mockParkingQuerySnapshot).documents }
-    verify(timeout(100)) { mockDocumentSnapshot.toObject(User::class.java) }
+    shadowOf(Looper.getMainLooper()).idle()
+
+    verify(mockDocumentReference, atLeast(1)).get()
   }
 
   @Test
-  fun getUserById_returnsCorrectValues() {
-    `when`(mockDocumentReference.get()).thenReturn(Tasks.forResult(mockDocumentSnapshot))
-    `when`(mockDocumentSnapshot.toObject(User::class.java)).thenReturn(user)
+  fun getUserById_callsOnFailure() {
+    val taskCompletionSource = TaskCompletionSource<DocumentSnapshot>()
+    `when`(mockDocumentReference.get()).thenReturn(taskCompletionSource.task)
 
     userRepositoryFirestore.getUserById(
         userId = user.public.userId,
-        onSuccess = { assert(it == user) },
-        onFailure = { fail("Expected success but got failure") })
-    verify(timeout(100)) { mockDocumentReference.get() }
-    verify(timeout(100)) { mockDocumentSnapshot.toObject(User::class.java) }
+        onSuccess = { fail("Expected failure but got success") },
+        onFailure = { assert(true) })
```

Comment:
Same here, the test is useless, if onFailure is not launched the test still passes

Reply:
If on failure is not launched then on sucess is? Is there a case where none is launched? Maybe I am confused...

Reply:
We can't assume that, this is the goal of these unit tests. To make sure we implemented it well and that there is no path that doesn't take any of the callbacks.

Also it's pretty hard to setup the mock correctly so that they call the callbacks.